extensions/notebook/src/book/bookTocManager.ts

```diff
@@ -0,0 +1,136 @@
+
+	constructor() {
+	}
+
+	async getAllFiles(toc: IJupyterBookSectionV2[], directory: string, filesInDir: string[]): Promise<IJupyterBookSectionV2[]> {
+		await Promise.all(filesInDir.map(async file => {
+			let isDirectory = (await fs.promises.stat(path.join(directory, file))).isDirectory();
+			if (isDirectory) {
+				let jupyterSection: IJupyterBookSectionV2 = {
+					title: file,
+					file: file,
+					expand_sections: true,
+					numbered: false,
+					sections: []
+				};
+				let files = await fs.promises.readdir(path.join(directory, file));
+				toc.push(jupyterSection);
+				await this.getAllFiles(toc, path.join(directory, file), files);
+			} else if (allowedFileExtensions.includes(path.extname(file))) {
+				const addFile: IJupyterBookSectionV2 = {
+					title: path.parse(file).name,
+					file: path.join(path.basename(directory), path.parse(file).name)
+				};
+				if (toc.length > 0) {
+					let indexToc = toc.findIndex(parent => parent.title === path.dirname(addFile.file));
+					if (indexToc !== -1) {
+						if (path.parse(addFile.file).ext === '.md') {
+							toc[indexToc].file = addFile.file;
+						} else {
+							toc[indexToc].sections.push(addFile);
+						}
+					}
+				} else if (path.parse(file).ext === '.md') {
+					//TODO : Add other types such as Intro, Introduction, Index etc etc
+					// what about two files named as Intro and then other as readme ?
+					let index: IJupyterBookSectionV2 = {
+						title: path.parse(file).name,
+						file: path.parse(file).name
+					};
+					toc.push(index);
+				}
+			}
+		}));
+		return toc;
+	}
+
+	hasSections(node: JupyterBookSection): boolean {
+		return (typeof node === 'object') && node.sections.length > 0 && (typeof node.sections !== 'undefined');
+	}
+
+	updateToc(tableOfContents: JupyterBookSection[], newToc: JupyterBookSection, findSection: JupyterBookSection[]): boolean {
+		if (tableOfContents === findSection) {
+			if (newToc !== undefined) {
+				tableOfContents.push(newToc);
+			} else {
+				//remove section
+				tableOfContents = [];
+			}
+			return true;
+		} else {
+			tableOfContents.forEach(t => this.hasSections(t) ? this.updateToc(t.sections, newToc, findSection) : undefined);
+		}
+		return false;
+	}
+
+	async createBook(bookContentPath: string, contentFolder: string): Promise<void> {
+		let filesinDir = await fs.promises.readdir(contentFolder);
+		let toc: IJupyterBookSectionV2[] = await this.getAllFiles([], contentFolder, filesinDir);
+		await fs.promises.mkdir(bookContentPath);
+		await fs.promises.writeFile(path.join(bookContentPath, '_config.yml'), yaml.safeDump({ title: path.basename(bookContentPath) }));
+		await fs.promises.writeFile(path.join(bookContentPath, '_toc.yml'), yaml.safeDump(toc, { lineWidth: Infinity }));
+		vscode.commands.executeCommand('notebook.command.openBook', bookContentPath, false);
+	}
+
+	async updateBook(element: BookTreeItem, updatedBook?: BookTreeItem): Promise<void> {
```

Comment:
Would be good to add xml doc comments for these - for example here I'm not really sure how "element" and "updatedBook" are related or what they correspond to

Reply:
I also think that trying to stuff a bunch of different types of "edits" into this single method is just making the whole thing a lot more confusing. Is there a reason you aren't making a separate method for each scenario such as `addSection`, `addNotebook`, etc?

(and I'll take time here again to reiterate how it would be utterly fantastic if we could clean up these models. None of it seems to make the clearest of sense - we keep passing around tree item objects for all sorts of non tree related tasks for example. And then we have books and notebooks sharing a model for some reason? Just imagine how much cleaner all this would be if we just had `NotebookModel` and `BookModel` and then could just have simple methods like I suggested above based on what type it is)

Reply:
I completely agree with separating the notebook logic from the book model. As well as having a version manager that handles all the related version operations and converts the jupyter book versions to our own version. 
I'll try to refactor as much as possible to simplify the logic.